src/ws/cert-session.c

```diff
@@ -0,0 +1,307 @@
+  char *authorization;
+  char *type = NULL;
+  const char *auth_value;
+  char **env;
+  int status;
+  int res;
+  int i;
+
+  program_name = basename (argv[0]);
+
+  if (isatty (0))
+    errx (2, "this command is not meant to be run from the console");
+
+  if (argc != 2)
+    errx (2, "invalid arguments to cockpit-cert-session");
+
+  /* Cleanup the umask */
+  umask (077);
+
+  rhost = getenv ("COCKPIT_REMOTE_PEER") ?: "";
+
+  save_environment ();
+
+  /* When setuid root, make sure our group is also root */
+  if (geteuid () == 0)
+    {
+      /* Always clear the environment */
+      if (clearenv () != 0)
+        err (1, "couldn't clear environment");
+
+      /* set a minimal environment */
+      setenv ("PATH", DEFAULT_PATH, 1);
+
+      if (setgid (0) != 0 || setuid (0) != 0)
+        err (1, "couldn't switch permissions correctly");
+    }
+
+  signal (SIGALRM, SIG_DFL);
+  signal (SIGQUIT, SIG_DFL);
+  signal (SIGTSTP, SIG_IGN);
+  signal (SIGHUP, SIG_IGN);
+  signal (SIGPIPE, SIG_IGN);
+
+  cockpit_authorize_logger (authorize_logger, DEBUG_SESSION);
+
+  /* Request authorization header */
+  write_authorize_begin ();
+  write_control_string ("challenge", "*");
+  write_control_end ();
+
+  /* And get back the authorization header */
+  authorization = read_authorize_response ("authorization");
+  auth_value = cockpit_authorize_type (authorization, &type);
+  if (!auth_value)
+    errx (EX, "invalid authorization header received");
+  if (strcmp (type, "tls-cert") != 0)
+    errx (2, "This authentication command only supports tls-cert");
+
+  pamh = perform_tlscert (rhost, auth_value);
+
+  cockpit_memory_clear (authorization, -1);
+  free (authorization);
+  free (type);
+
+  for (i = 0; env_saved[i] != NULL; i++)
+    pam_putenv (pamh, env_saved[i]);
+
+  env = pam_getenvlist (pamh);
+  if (env == NULL)
+    errx (EX, "get pam environment failed");
+
+  if (want_session)
+    {
+      assert (pwd != NULL);
+
+      if (initgroups (pwd->pw_name, pwd->pw_gid) < 0)
+        err (EX, "%s: can't init groups", pwd->pw_name);
+
+      signal (SIGTERM, pass_to_child);
+      signal (SIGINT, pass_to_child);
+      signal (SIGQUIT, pass_to_child);
+
+      utmp_log (1, rhost);
+
+      status = fork_session (env, session);
+
+      utmp_log (0, rhost);
+
+      signal (SIGTERM, SIG_DFL);
+      signal (SIGINT, SIG_DFL);
+      signal (SIGQUIT, SIG_DFL);
+
+      res = pam_setcred (pamh, PAM_DELETE_CRED);
+      if (res != PAM_SUCCESS)
+        err (EX, "%s: couldn't delete creds: %s", pwd->pw_name, pam_strerror (pamh, res));
+      res = pam_close_session (pamh, 0);
+      if (res != PAM_SUCCESS)
+        err (EX, "%s: couldn't close session: %s", pwd->pw_name, pam_strerror (pamh, res));
+    }
+  else
+    {
+      status = session (env);
+    }
+
+  pam_end (pamh, PAM_SUCCESS);
+
+  free (last_err_msg);
+  last_err_msg = NULL;
+  free (last_txt_msg);
+  last_txt_msg = NULL;
+  free (conversation);
+  conversation = NULL;
```

Comment:
The  `main` function is mostly a copy&paste of session.c; but as we use g_auto pointers in the new code above, they should also be used for main for consistency.

Reply:
Nevermind, these are global variables.